File: api/functions/release-detail.ts
// API endpoint: GET /api/release/{artist}/{release}
//
// One release's buying guide as JSON — the twin of the HTML page `api/edge/release-page.ts`
// serves at /a/{artist}/{release}. It exists so the Mac app and the browser extension can
// render the payout comparison natively instead of doing what they do today, which is
// `NSWorkspace.shared.open(url)` / `chrome.tabs.create` and handing the whole point of the
// product to a browser tab.
//
// **Not a second renderer of that URL.** The UNS-100 bifurcation rule is about one URL with two
// renderers; this is a different URL returning data, and the HTML page remains the only thing
// that renders /a/{artist}/{release}. What the two must share is the *answer*: both run the same
// query (`getReleaseDetail`) and apply the same payout rules, so a native client can never
// describe a release differently from the web page a fan would see for it.
//
// **`payoutPercent` is computed here, per source, on purpose.** Payout figures are duplicated by
// hand across eight files in this repo, and that drift is what let the Discord bot quote an
// unsourced '86-95%' for Jam.coop to users for months (fixed in PR #389). A client that reads
// the number off the response cannot drift. It also can't miss the Bandcamp Friday override
// below, which no client currently knows exists.

import { getReleaseDetail, type ReleaseDetailSource } from './db';
import { checkRateLimit, getClientIp } from './ratelimit';
import { Sentry } from '../lib/sentry';
import { PLATFORMS } from '../shared/platform-registry';
import { isBandcampFriday } from '../shared/bandcamp-friday';
import { AVAILABILITY_ORDER, payoutRank } from '../shared/release-display';

const CORS_HEADERS: Record<string, string> = {
  'Content-Type': 'application/json',
  // The wildcard is deliberate, not an oversight — the same reason check-releases.ts carries
  // one: the Mac app and the browser extension both call this endpoint, and neither sends an
  // Origin that the shared buildCorsHeaders() allowlist (unstream.stream) would accept. The
  // body is the same public data the release page already serves to anyone.
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'GET, OPTIONS',
  'Access-Control-Allow-Headers': 'Content-Type',
};

/**
 * Both slug generators — `artistSlug()` and `releaseSlug()` in release-utils.ts — emit only
 * lowercase alphanumerics and hyphens, so anything else cannot be a slug we ever minted.
 * Refusing those before touching the database means a scan of junk paths costs nothing.
 *
 * Checked against production: all 842 `releases.slug` values match, and the only one of 3,416
 * `artists.slug` values that doesn't is a single empty string — which no URL can address anyway
 * (`/api/release//x` doesn't match the route, and the empty-param check above catches it).
 */
const SLUG_PATTERN = /^[a-z0-9][a-z0-9-]{0,119}$/;

function json(statusCode: number, body: unknown, extraHeaders: Record<string, string> = {}) {
  return { statusCode, headers: { ...CORS_HEADERS, ...extraHeaders }, body: JSON.stringify(body) };
}

/**
 * The two slugs, from the request path.
 *
 * **`rawUrl`, not `event.path`.** Behind a `status = 200` rewrite `event.path` can be the rewrite
 * *target* (`/.netlify/functions/release-detail`), which carries no routing information at all.
 * Netlify always sets `rawUrl` to the full original request URL, so that is the source of truth —
 * the same arrangement feed-releases.ts uses, for the same reason.
 *
 * This originally read the slugs from `queryStringParameters`, on the assumption that Netlify
 * substitutes `from` placeholders into a rewrite destination's query string. It does not: on
 * deploy preview 393 the rewrite matched and this function ran with an empty
 * `queryStringParameters`, so every request 400'd. The query-string branch is kept only as a
 * fallback for direct invocation — `/.netlify/functions/release-detail?artist=…&release=…` is a
 * real, working way to call this, and the tests drive it that way too.
 */
export function parseSlugs(event: {
  path?: string;
  rawUrl?: string;
  queryStringParameters?: Record<string, string> | null;
}): { artist: string; release: string } | null {
  const path = (() => {
    if (event.rawUrl) {
      try {
        return new URL(event.rawUrl).pathname;
      } catch {
        // fall through to `path`
      }
    }
    return event.path || '';
  })();

  const match = path.replace(/\/$/, '').match(/^\/api\/release\/([^/]+)\/([^/]+)$/);
  if (match) {
    try {
      return { artist: decodeURIComponent(match[1]), release: decodeURIComponent(match[2]) };
    } catch {
      // A malformed percent-escape can't be a slug we minted; fall through and let the caller 404.
      return null;
    }
  }

  const artist = event.queryStringParameters?.artist;
  const release = event.queryStringParameters?.release;
  return artist && release ? { artist, release } : null;
}

/** One source's offers, ordered the way the release page orders them: what you can buy first,
 *  cheapest first within that, and anything you can't buy at the bottom. */
function sortedOffers(source: ReleaseDetailSource) {
  return [...source.offers].sort((a, b) => {
    const availability = (AVAILABILITY_ORDER[a.availability] ?? 2) - (AVAILABILITY_ORDER[b.availability] ?? 2);
    if (availability !== 0) return availability;
    return (a.price ?? Infinity) - (b.price ?? Infinity);
  });
}

export async function handler(event: {
  httpMethod?: string;
  path?: string;
  rawUrl?: string;
  queryStringParameters?: Record<string, string> | null;
  headers?: Record<string, string | undefined>;
}) {
  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 204, headers: CORS_HEADERS, body: '' };
  }
  if (event.httpMethod && event.httpMethod !== 'GET') {
    return json(405, { error: 'Method not allowed' });
  }

  // 'lenient' because a client fetches this when a person taps a release, and a household
  // behind one IP browsing a few records is normal traffic rather than abuse.
  const rl = await checkRateLimit(getClientIp(event.headers || {}), 'lenient', CORS_HEADERS);
  // `response` is optional on the result type, so it gets a 429 of its own rather than a cast:
  // returning `undefined` here would serve a rate-limited request as an empty 200.
  if (rl.limited) return rl.response ?? json(429, { error: 'Rate limit exceeded' });

  const slugs = parseSlugs(event);
  if (!slugs) {
    return json(400, { error: 'artist and release are required' });
  }
  const { artist, release } = slugs;

  if (!SLUG_PATTERN.test(artist) || !SLUG_PATTERN.test(release)) {
    return json(404, { error: 'Release not found' });
  }

  try {
    const { detail, failed } = await getReleaseDetail(artist, release);

    // The lookup itself broke. A 503 with no caching, never a 404 — a Supabase outage rendered
    // as "this release doesn't exist" would be a convincing lie, and one the CDN would then
    // hold for five minutes. Never cache uncertainty.
    if (failed) {
      Sentry.captureMessage('[release-detail] release lookup failed', {
        level: 'error',
        tags: { artist, release },
        extra: { context: 'release-detail.lookupFailed' },
      });
      return json(503, { error: 'Release lookup temporarily unavailable' }, { 'Cache-Control': 'no-store' });
    }

    // Genuinely absent: an unknown artist, a stale link from an old alert, or a release the
    // artist has since hidden. Deliberately one 404 for all of them — `is_hidden` exists to
    // make a suppressed release indistinguishable from one that was never catalogued.
    if (!detail) {
      return json(404, { error: 'Release not found' });
    }

    const bandcampFriday = isBandcampFriday();

    // Artist-paying first, the same ordering the release page and the feeds use. Without it a
    // secondhand Discogs listing could lead over a direct purchase from the artist, which would
    // be off-mission even though both facts are true.
    const sources = [...detail.release.sources]
      .sort((a, b) => payoutRank(b.platform) - payoutRank(a.platform))
      .map(source => {
        const info = PLATFORMS[source.platform];
        const isBCFriday = source.platform === 'bandcamp' && bandcampFriday;
        // On a Bandcamp Friday Bandcamp waives its revenue share, so the registry's usual range
        // is simply wrong for the next 24 hours. The release page already overrides it; a client
        // rendering the same guide from this response has to get the same number.
        return {
          platform: source.platform,
          name: info?.name ?? source.platform,
          url: source.url,
          payoutPercent: isBCFriday ? '~97%' : (info?.payoutPercent ?? null),
          bandcampFriday: isBCFriday,
          detailCheckedAt: source.detailCheckedAt,
          offers: sortedOffers(source),
        };
      });

    // The oldest price across every source is the honest freshness claim: saying "checked today"
    // because *one* source was re-read would overstate the rest. ISO rather than the page's
    // "3 days ago" — a native client formats dates in the user's own locale.
    const pricesCheckedAt = sources
      .flatMap(s => s.offers.map(o => o.capturedAt))
      .filter(Boolean)
      .sort()[0] ?? null;

    return {
      statusCode: 200,
      headers: {
        ...CORS_HEADERS,
        'Cache-Control': 'public, max-age=0, must-revalidate',
        'Netlify-CDN-Cache-Control': 's-maxage=300, stale-while-revalidate=86400',
        'Cache-Tag': `release-detail-${artist}-${release}`,
      },
      body: JSON.stringify({
        artist: detail.artist,
        release: {
          slug: detail.release.slug,
          title: detail.release.title,
          releaseType: detail.release.releaseType,
          releaseDate: detail.release.releaseDate,
          datePrecision: detail.release.datePrecision,
          status: detail.release.status,
          artworkUrl: detail.release.artworkUrl,
          pricesCheckedAt,
          sources,
        },
        // The page URL, so a native client can offer "open the full guide" without rebuilding
        // the URL shape — and so the two surfaces stay tied together if it ever changes.
        pageUrl: `https://unstream.stream/a/${encodeURIComponent(artist)}/${encodeURIComponent(release)}`,
        bandcampFriday,
      }),
    };
  } catch (error) {
    console.error('[release-detail] Error:', error);
    return json(500, { error: 'Internal server error' });
  }
}

File: apps/mac/Unstream/Models/ReleaseDetail.swift
import Foundation

// One release's buying guide, as served by GET /api/release/{artist}/{release}.
//
// The payout comparison is the whole point of the Releases feature, and until now both clients
// handed that moment to a browser tab. These types are what let the app show it itself.
//
// **Payout percentages come off the wire, never out of `platformCatalog`.** The registry figures
// are duplicated by hand across eight files in this repo, and that drift is what let the Discord
// bot quote an unsourced rate for Jam.coop for months. A client that reads the number the server
// computed cannot drift — and it also gets the Bandcamp Friday override (~97%), which no client
// knows exists.

/// One buyable thing: a format, at a price, in a state of availability.
struct ReleaseDetailOffer: Codable, Hashable, Identifiable {
    let format: String
    let price: Double?
    let currency: String?
    let availability: String
    let capturedAt: String?

    var id: String {
        let priceKey = price.map { String($0) } ?? "none"
        return "\(format)-\(priceKey)-\(currency ?? "")-\(availability)"
    }

    /// Sold out is the only state that means "you cannot give this artist money right now".
    /// `unknown` still gets a Buy button, because the platform page may well sell it.
    var isBuyable: Bool { availability != "sold_out" }
}

/// One platform selling the release.
struct ReleaseDetailSource: Codable, Hashable, Identifiable {
    let platform: String
    /// The platform's proper display name, from the server's registry — "Jam.coop", not "Jamcoop".
    let name: String
    let url: String
    /// A range like "80-85%", or nil for a platform with no published rate (Discogs listings are
    /// secondhand, so the artist genuinely receives nothing — worth saying, not worth hiding).
    let payoutPercent: String?
    /// True while Bandcamp is waiving its revenue share, which makes `payoutPercent` ~97%.
    let bandcampFriday: Bool
    /// When this platform's page was last read for prices. Null means never — which is not the
    /// same as "it sells nothing", and the UI has to keep those apart.
    let detailCheckedAt: String?
    let offers: [ReleaseDetailOffer]

    var id: String { platform }
}

struct ReleaseDetailRelease: Codable, Hashable {
    let slug: String
    let title: String
    let releaseType: String
    let releaseDate: String?
    let datePrecision: String?
    let status: String
    let artworkUrl: String?
    /// The oldest price across every source — the honest freshness claim, ISO-8601.
    let pricesCheckedAt: String?
    /// Already ordered artist-paying-first by the server, and each source's offers already
    /// ordered buyable-and-cheapest-first. **Do not re-sort.** Ordering needs the payout
    /// registry, and duplicating that here is exactly the drift these types exist to avoid.
    let sources: [ReleaseDetailSource]
}

struct ReleaseDetailArtist: Codable, Hashable {
    let slug: String
    let name: String
    let imageUrl: String?
}

struct ReleaseDetail: Codable, Hashable {
    let artist: ReleaseDetailArtist
    let release: ReleaseDetailRelease
    /// The web release page for this record, for the "Open on Unstream" escape hatch. Taken from
    /// the response rather than rebuilt, so the two surfaces stay tied together if it ever changes.
    let pageUrl: String
    let bandcampFriday: Bool

    var isUpcoming: Bool { release.status == "announced" }
}

// MARK: - Formatting
//
// Mirrors api/shared/release-display.ts. Deliberately small and in one place: every function here
// makes a claim about someone's money or a release date, so the rules must not drift between the
// web page and the app. `UnstreamTests/ReleaseDetailTests.swift` pins each of them.

enum ReleaseFormatting {
    static let formatLabels: [String: String] = [
        "digital": "Digital", "vinyl": "Vinyl", "cassette": "Cassette",
        "cd": "CD", "book": "Book", "merch": "Merch", "other": "Other",
    ]

    /// Only the states worth saying out loud — "available" needs no label beside a price.
    static let availabilityLabels: [String: String] = [
        "preorder": "Pre-order", "sold_out": "Sold out", "unknown": "Price unknown",
    ]

    static func label(forFormat format: String) -> String {
        formatLabels[format] ?? format.capitalized
    }

    /// Money, in the currency the platform quoted.
    ///
    /// Two decimals or none — never one. "$25", never "$25.00"; "$8.50", never "$8.5", which
    /// looks like a typo in a price. Cents are never rounded away: tidying someone's money is a
    /// small lie in an app whose entire job is being accurate about it.
    ///
    /// `en_US` rather than the user's locale, deliberately. This is the one place where matching
    /// the web page matters more than feeling native — the endpoint exists so a native client
    /// can never describe a release differently from the page a fan would see for it, and the
    /// page formats with `Intl.NumberFormat('en-US', …)`.
    static func money(_ amount: Double, _ currency: String?) -> String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = currency ?? "USD"
        formatter.locale = Locale(identifier: "en_US")
        let isWhole = amount.rounded() == amount
        formatter.minimumFractionDigits = isWhole ? 0 : 2
        formatter.maximumFractionDigits = isWhole ? 0 : 2
        return formatter.string(from: NSNumber(value: amount)) ?? "\(amount) \(currency ?? "")"
    }

    /// What goes in the price column.
    ///
    /// **Zero means name-your-price, not free.** Bandcamp reports a name-your-price release as
    /// `price: 0` with no other signal, and rendering that as "$0" tells a fan the record costs
    /// nothing — when they are in fact being invited to decide what to pay, which in an app about
    /// paying artists is close to the opposite message. Caught on Kid Lightbulbs' own catalog,
    /// where every release is name-your-price.
    ///
    /// A nil price is a format we know exists but have no figure for; an em dash says that
    /// without claiming it's free.
    static func price(_ price: Double?, _ currency: String?) -> String {
        guard let price else { return "—" }
        if price == 0 { return "Name your price" }
        return money(price, currency)
    }

    /// "≈$4–$4.25 to artist" — the emotional payload of the whole product, at the moment someone
    /// is deciding where to buy.
    ///
    /// Deliberately a **range**, because the payout figures are ranges and they are honest about
    /// it: Bandcamp's real take differs between digital and physical, payment processing comes off
    /// the top, and on a Bandcamp Friday it's ~97%. Asserting a single precise figure about
    /// someone's income would be worse than saying "roughly". Returns nil rather than inventing
    /// either a percentage or a price.
    static func payout(_ price: Double?, _ currency: String?, _ payoutPercent: String?) -> String? {
        guard let price, price > 0, let payoutPercent else { return nil }

        let numbers = payoutPercent
            .components(separatedBy: CharacterSet(charactersIn: "0123456789.").inverted)
            .filter { !$0.isEmpty }
            .compactMap(Double.init)
        guard let low = numbers.first, let high = numbers.last else { return nil }

        let lowAmount = money(price * low / 100, currency)
        let highAmount = money(price * high / 100, currency)
        return lowAmount == highAmount
            ? "≈\(lowAmount) to artist"
            : "≈\(lowAmount)–\(highAmount) to artist"
    }

    /// A release date rendered only as precisely as we actually know it.
    ///
    /// MusicBrainz returns year-only and month-only dates, and printing "1 January 2023" for a
    /// year-only date states a fact no source ever gave us. `datePrecision` exists for this.
    ///
    /// Unlike money, the *shape* of a date is safe to localize — "23 October 2025" and
    /// "October 23, 2025" are the same claim — so the template follows the user's locale.
    static func date(_ iso: String?, precision: String?) -> String {
        guard let iso, let year = Int(iso.prefix(4)) else { return "" }
        if precision == "year" { return String(year) }
        guard precision == "month" || precision == "day" else {
            // 'unknown' precision on a stored date means we padded it ourselves. Say nothing
            // rather than pick a shape.
            return ""
        }

        let parts = iso.split(separator: "-").compactMap { Int($0) }
        var components = DateComponents()
        components.year = year
        components.month = parts.count > 1 ? parts[1] : 1
        components.day = parts.count > 2 ? parts[2] : 1
        guard let date = Calendar(identifier: .gregorian).date(from: components) else {
            return String(year)
        }

        let formatter = DateFormatter()
        formatter.setLocalizedDateFormatFromTemplate(precision == "month" ? "MMMMy" : "dMMMMy")
        return formatter.string(from: date)
    }

    /// "checked today" / "checked 3 days ago" — how old a price is, in words a person reads.
    /// Empty when we've never read prices, so the caller can say that instead of implying zero.
    ///
    /// Whole elapsed 24-hour periods, matching `relativeDays` in api/shared/release-display.ts
    /// exactly. Deliberately not `Calendar.dateComponents([.day])`, which counts in the user's
    /// local calendar: a price captured at 23:00 UTC would read "yesterday" to a fan in one
    /// timezone and "today" to a fan in another, for the same release at the same moment.
    static func freshness(_ iso: String?, now: Date = Date()) -> String {
        guard let iso, let captured = isoDate(iso) else { return "" }
        let days = Int(floor(now.timeIntervalSince(captured) / 86_400))
        if days < 0 { return "Prices checked just now" }
        if days == 0 { return "Prices checked today" }
        if days == 1 { return "Prices checked yesterday" }
        return "Prices checked \(days) days ago"
    }

    /// The API sends fractional seconds on some timestamps and not others, and
    /// `ISO8601DateFormatter` fails the whole parse on the wrong option rather than coping.
    private static func isoDate(_ iso: String) -> Date? {
        let withFractional = ISO8601DateFormatter()
        withFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = withFractional.date(from: iso) { return date }
        return ISO8601DateFormatter().date(from: iso)
    }
}
